scene/main/canvas_item.cpp

```diff
@@ -557,38 +557,53 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
 		}
 
 		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, p_rect, p_color);
+	} else if (p_width >= p_rect.size.width || p_width >= p_rect.size.height) {
+		Size2 pos = p_rect.position - Size2(0.5 * p_width, 0.5 * p_width);
+		Size2 size = p_rect.size + Size2(p_width, p_width);
+		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, Rect2(pos, size), p_color);
 	} else {
-		// Thick lines are offset depending on their width to avoid partial overlapping.
-		// Thin lines don't require an offset, so don't apply one in this case
-		real_t offset;
-		if (p_width >= 2) {
-			offset = p_width / 2.0;
-		} else {
-			offset = 0.0;
+		Size2 pos = p_rect.position;
+		Size2 size = p_rect.size;
+
+		// Snap to pixel grid for odd `p_width`.
+		if (Math::is_equal_approx(Math::fmod(p_width, 2), 1)) {
+			pos += Size2(0.5, 0.5);
+			size -= Size2(1, 1);
```

Comment:
Again, I don't get what's the logic behind that kind of strange cases. It does break things when the condition is satisfied.

<details>
<summary>Example:</summary>

![ArXQ3IhBR5](https://user-images.githubusercontent.com/9283098/205374178-f1e5fef7-8cf8-41f8-9eba-4631fd67d0e8.png)
![godot windows editor x86_64_2yO3J9qvS4](https://user-images.githubusercontent.com/9283098/205374173-4954028c-ca36-4c49-86f1-e7daa920faad.png)
![8g6mbuTXmT](https://user-images.githubusercontent.com/9283098/205374181-ecb4a3aa-798b-4d7e-b3cd-8849f41f9fce.png)

</details>

What issue is it attempting to solve?

Reply:
![](https://user-images.githubusercontent.com/47700418/205376111-d62b29d2-4408-4ab7-8356-3c4017f3ec6a.png)

Reply:
Maybe we could add an optional `pixel_perfect` parameter?

Reply:
I changed my mind. This point should be explained in the documentation/tutorial, and then the user can adjust the position and size for odd widths when calling `draw_rect` if they wish.